src/lib/Lib/CMakeExecution.cpp

```diff
@@ -0,0 +1,164 @@
+    int const result = llvm::sys::ExecuteAndWait(cmakePath, args, emptyEnv, redirects);
+    MRDOCS_CHECK(result == 0, "CMake execution failed");
+
+    auto const bufferOrError = llvm::MemoryBuffer::getFile(outputPath);
+    MRDOCS_CHECK(bufferOrError, "Failed to read CMake help output");
+
+    return bufferOrError.get()->getBuffer().str();
+}
+
+Expected<std::string>
+getCmakeDefaultGenerator(llvm::StringRef cmakePath) 
+{
+    MRDOCS_TRY(auto const cmakeHelp, executeCmakeHelp(cmakePath));
+
+    std::istringstream stream(cmakeHelp);
+    std::string line;
+    std::string defaultGenerator;
+
+    while (std::getline(stream, line)) {
+        if (line[0] == '*' && line[1] == ' ') {
+            size_t const start = 2;
+            size_t const end = line.find("=", start);
+            if (end == std::string::npos) {
+                continue;
+            }
+            return line.substr(start, end - start);
+        }
+    }
+    return Unexpected(Error("Default CMake generator not found"));
+}
+
+Expected<bool>
+cmakeDefaultGeneratorIsVisualStudio(llvm::StringRef cmakePath) 
+{
+    MRDOCS_TRY(auto const defaultGenerator, getCmakeDefaultGenerator(cmakePath));
+    return defaultGenerator.starts_with("Visual Studio");
+}
+
+std::vector<std::string>
+parseCmakeArgs(std::string const& cmakeArgsStr) 
+{
+    std::vector<std::string> parsedArgs;
+    std::string arg;
+    bool inQuotes = false;
+
+    for (char ch : cmakeArgsStr) 
+    {
+        if (ch == '"') 
+        {
+            inQuotes = !inQuotes;
+        } 
+        else if (std::isspace(ch) && !inQuotes) 
+        {
+            if ( ! arg.empty()) 
+            {
+                if (arg[0] == '-' && arg.size() == 2) 
+                    continue;
+                parsedArgs.push_back(arg);
+                arg.clear();                
+            }
+        } 
+        else 
+        {
+            arg += ch;
+        }
+    }
+
+    if ( ! arg.empty()) 
+    {
+        parsedArgs.push_back(arg);
+    }
+
+    return parsedArgs;
+}
+
+} // anonymous namespace
+
+Expected<std::string>
+executeCmakeExportCompileCommands(llvm::StringRef projectPath, llvm::StringRef cmakeArgs) 
+{
+    MRDOCS_TRY(auto const cmakePath, getCmakePath());
+    MRDOCS_CHECK(llvm::sys::fs::exists(projectPath), "CMakeLists.txt not found");
+
+    llvm::SmallString<128> tempDir;
+    MRDOCS_CHECK(!llvm::sys::fs::createUniqueDirectory("compile_commands", tempDir), "Failed to create temporary directory");
+
+    std::optional<llvm::StringRef> const redirects[] = {llvm::StringRef(), llvm::StringRef(), llvm::StringRef()};
+    std::vector<llvm::StringRef> args = {cmakePath, "-S", projectPath, "-B", tempDir.str(), "-DCMAKE_EXPORT_COMPILE_COMMANDS=ON"};
+
+    auto const additionalArgs = parseCmakeArgs(cmakeArgs.str());
+    bool visualStudioFound = false;
+    for (auto const& arg : additionalArgs) 
+    {
+        if (arg.starts_with("-G") && arg.find("Visual Studio", 2) != std::string::npos) 
```

Comment:
Oh... so you put keys and values in the same arg to make it easier to find here. Not a good idea in terms of separation of concerns. Except for -D arguments, other arguments do need to be separated and the quotes can't be mixed when the API requotes them. The case where the key is -G can always be easily handled separately (if -G then check if continuation or next arg starts with Visual Studio).

Reply:
I'm combining the arguments (like `-D KEY=VALUE` or `-G VALUE`) because when passed as individual elements in the vector to `llvm::sys::ExecuteAndWait()`, each argument is sent to CMake in a way similar to what you described over Slack. As per your suggestion of formatting arguments like `"-DA=B C"`, keeping them combined ensures that they're processed correctly by CMake. While it might not be optimal in terms of software design purity, practically, it works well. I've tested various ways of specifying quotes (realized I missed escaping them, thanks) and this approach seems to function effectively. This piece of code is not intended to be a comprehensive CMake argument parser but rather to address our specific use case. 
What do you think?

Reply:
Additionally, I'm aware that if the code starts to grow and we need to add more special conditions for argument parsing, it might become unwieldy or less clean. In that scenario, I agree that a refactoring of the parser would be necessary to maintain clarity and manageability. But as of now, it seems to serve our practical purposes well without overcomplicating things.

Reply:
Oh... I see.

> As per your suggestion of formatting arguments like "-DA=B C", keeping them combined ensures that they're processed correctly by CMake

I think I miscommunicated that then. The comment was about args with literal `\"`s in it. Not about all combinations of key args (starting with `-`) and value args. 

For args in general, you just push them to the list of args as usual. There's no need to merge `-D` with the following arg. The logic you implemented is not only for `-D` but for all keys, and this merged arg will not always be valid. So you can just implement the simpler logic here.

So, about the point I was making, if you choose to implement it, is all about quotations. The problem is bash recognizes quotations around args (for instance, `"A C"` to represent a single arg) while cmake recognizes quotations inside args (for instance, `D="A C"` to represent a single pair). Note that make understands `D="A C"` as a single arg representing a pair while bash doesn't understand that. Bash understands `D="A C"` to be 2 args: `D="A` and `C"`. CMake uses two passes to convert the arguments bash understands to the arguments it understands. 

The problem this causes is that if your API to call processes attempt to quote everything again, when sending `D="A C"` to CMake, it will send something like `"D=\"A"` and `"C\""`, and CMake will not be able to reconstruct `D="A C"` from that. First you'd have to check if this is really a problem because the LLVM API for calling the process might not have this problem. Second, if this is a problem, you might not choose to fix it for now. Third, if you choose to fix the problem, I don't know what the best solution for this is but what I implemented in cpp-actions was logic to convert all args like `D="A C"` to `"D=A C"`, which is recognizable both by bash and cmake. [This](https://github.com/alandefreitas/cpp-actions/blob/89228895355dc399f2e1ef9d310c5f9695e30183/cmake-workflow/index.js#L1090) is the specific regex I used but, as I said, I'm not even sure that's the best solution.

> This piece of code is not intended to be a comprehensive CMake argument parser but rather to address our specific use case.

To be clear, the problem with that logic is not that it's incomplete. It's that it includes more complex logic that was effectively making things worse. Namely, your algorithm is joining any arg with the preceding arg if the preceding arg whenever it starts with `-`. At best, this makes things unnecessarily more complex. At worst, it breaks things because this is not valid for all args.

This is the main issue. The problem with arguments that contain quotes is an independent problem.